### PR TITLE
PAID Proposal: Solido Client RPC

### PR DESCRIPTION
**Pull request by molekilla**
# Grant Application Checklist

- [x ] The [application-template.md](https://github.com/w3f/Open-Grants-Program/blob/master/applications/application-template.md) has been copied, renamed ( "solido.md") and updated.
- [ x] A BTC address for the payment of the milestones is provided inside the application.  
- [x ] The software of the project will be released under the Apache license version 2.0 as specified in the [terms and conditions](https://gist.github.com/Noc2/75bc58e8ce9b5d419ff883b0cf2b8c19).
- [ x] The total funding amount of the project is below CHF 30k at the time of submission.
- [ x] The initial PR contains only one commit (squash if needed before submitting your PR).

### CODE REVIEW

**Comment by molekilla** [comment]
@mmagician hi we are awaiting feedback let us know, thans

**Review by mmagician — CHANGES_REQUESTED** [review]
Thanks for creating a new application. As I've never worked with GraphQL, I am finding it hard to imagine a use-case scenario or an example of how it would work on Polkadot. So if I understand correctly, for each publicly callable functions, and for each pallet that a chain integrates, you will provide a corresponding GraphQL API method?

How would developers benefit from such a mapping, i.e. what do they need to do on their end in order to use e.g. Plasm's callables in their code? Do they need to import the TypeScript module that is part of your deliverables?

It looks you want to start with Plasm - could you perhaps provide an example of API subset for that chain, i.e. how you envision it would look like?

Also, is the mapping of chain's callables & query endpoints to GraphQL API done manually per each method, or is it possible to automate this? I guess for one specific chain it's OK to do manually (though still plenty of mappings!) - but it would be quite a workload for each parachain.

### ----------------

> management of schema and mutations, using a single client to multiple GraphQL API's.

Please explain in more detail the above point.

### ----------------
> integrated into a wallet 

So the wallet would use GraphQL APIs directly? Does that mean you're developing your own wallet? 

### ----------------
I hope my questions make sense, else please forgive my ignorance if I have misunderstood your application.

**Comment by mmagician** [comment]
Thanks for the detailed answer, you seem a motivated team indeed.
1. > the idea is that each Network [...] has an interface within our client

But who's creating this interface? You haven't really answered my question on whether the mapping (i.e. interface creation) is done manually or can be auto-generated - and who's responsible for this: you as the client creator, or the network maintainers?

2. I think I get an idea now of what you're trying to accomplish. I feel like polkadot-js already exposes the pallet interface to the developers quite nicely: https://github.com/polkadot-js/api/tree/master/packages/api. Is this the kind of simplification that your TypeScript module would provide?

For example, you can write:
```
const { ApiPromise, WsProvider } = require('@polkadot/api');
const api = await ApiPromise.create({ provider: new WsProvider('ws://127.0.0.1:9944') }); 
const result = await api.tx.<module_name>.<method_name>(<arguments>).signAndSend(...)
```

**Comment by alfredolopez80** [comment]
Thanks, and Really we are soo motivated to deploy this solution and integrated into PolkaDOT/Substrate.io.

## About your Question

### But who's creating this interface? You haven't really answered my question on whether the mapping (i.e. interface creation) is done manually or can be auto-generated - and who's responsible for this: you as the client creator, or the network maintainers?

Yes, we create these connectors at least for EVM Pallet, Bitcoin Parachain (when is online), and of course for PolkaDOT, Plasm, and moonbeam, and after that, we checking with the community the new needs about other parachain's and develop those connectors

### Point 2 (I think I get an idea now of what you're trying to accomplish. I feel like polkadot-js already exposes the pallet interface to the developers quite nicely: https://github.com/polkadot-js/api/tree/master/packages/api. Is this the kind of simplification that your TypeScript module would provide?)

Said library only simplifies access to polkaDOT RPCs as such, our client intends to go much further than I have inspected as I mentioned in lunar.dev and Rosetta-SDK, the idea is that said API not only allows simple and standard access to multiple parachain's within polkaDOT (which in part Polkadot-JS API), but will additionally allow it to be integrated into a GraphQL engine that will allow more elegant handling of the data that is received from the RPC methods, and additional handling queries and mutations of said data to the front-end without requiring a server or intermediate API's, but will be done directly in the Dapp, which greatly simplifies the development at the FrontEnd level, in addition to simplifying the infrastructure required for the development of Decentralized Applications (Dapp's)

Additionally, the idea is that the handling of parameters required to configure the connectors is as simple as possible, in addition to avoiding redundancy, and that the RPC methods can be called statically in case of the standard methods, and dynamically in case of particular methods of the parachain, all this through a syntax based on TypeScript, and with the ability to handle observables integrated to the connectors based on GraphQL.

As a final comment, I consider you should delve into the advantages that GraphQL offers, in data management at the frontend level, it is a technology developed by the ReactJS / Facebook Team, for handling large volumes of data quickly and very agilely, in addition to allowing data mutations in a very practical way, almost immediately after obtaining the result of the queries, this is used by Facebook, so as not to handle all the volume of data that they may have in their database about a user, but only abstract a part of it, and this is filtered and treated according to the needs of the view (at the front level) of the user in question, giving Dapps a power never seen before, especially in case of high data volumes. . In addition to being a tool developed to work with Data Stream, which is normal when you interact with a blockchain.

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks for the application. Sounds like an interesting project. Are you aware of [Substrate Graph](https://github.com/playzero/substrate-graph) and especially [Hydra](https://github.com/Joystream/hydra). How is your solution different? And generally what’s your focus? (smart contracts in ink! or runtimes, validators or users, etc.) Or do you want to support everything?

**Comment by rubengue** [comment]
The original Solido codebase was centered around an EVM ORM like abstraction. Back in those days, our experience was working with VEchain, which has two differents clients technologies: A Web3 extension call "thorify" and a propietary client RPC technology call "connex" 

```javascript

import { Framework } from '@vechain/connex-framework';
import { DriverNodeJS } from '@vechain/connex.driver-nodejs';

import {
  SolidoModule,
} from '@decent-bet/solido';
import {
 ConnexPlugin,
 ConnexSettings,
} from '@decent-bet/solido-provider-connex';
import {
 ThorifyPlugin,
 ThorifySettings,
} from '@decent-bet/solido-provider-thorify';
import { EnergyTokenContract, EnergyContractImport } from './EnergyContract';
import Web3 from 'web3';
const { thorify } = require('thorify');

// Create Solido Module
export const module = new SolidoModule(
  [
    {
      name: 'ConnexToken',
      import: EnergyContractImport,
      entity: EnergyTokenContract,
      provider: ConnexPlugin,
    },
    {
      name: 'ThorifyToken',
      import: EnergyContractImport,
      entity: EnergyTokenContract,
      enableDynamicStubs: true,
      provider: ThorifyPlugin,
    }
  ],
);

const privateKey = '0x............';
const chainTag = '0x4a';
const defaultAccount = '0x...........';
const thorUrl = 'http://localhost:8669';

const thor = thorify(new Web3(), thorUrl);

const contracts = module.bindContracts();
const token = contracts.getContract<EnergyTokenContract>('ThorifyToken');
token.onReady<ThorifySettings>({
  privateKey,
  thor,
  defaultAccount,
  chainTag
});

(async () => {
  const balance = await token.balanceOf(defaultAccount);
  console.log(balance);
})();

```

Back to 2021, we think that the industry has strong preferences for GraphQL RPC client because:

* The Graph has had success using GraphQL in their indexing protocol. 
* Ethereum V1 added a year ago or so GraphQL support in geth.
* web3api.dev is also exploring using GraphQL as a client RPC.

To summarize, what make us different is that, instead of recreating everything from scratch, we found a better strategy to tackle this problem: 

* We map implementation from GraphQL to OpenApi using an open source connector that translates Rosetta OpenApi to GraphQL
* Thus, we can query and mutate using a single GraphQL query, one or more GraphQL model namespaces, without having to deal with specific RPC languages 

```javascript

{
  queryChains(
      chains: ["plasm", "moonbeam", "ethereum", "vechain"],
    abiSignatures: [
      "event LogSetQuestOutcome(bytes32 indexed id,address indexed user,uint256 questEntryCount)",
      "event LogPayForQuest(bytes32 indexed id,address indexed user, address indexed payer,uint256 questEntryCount)"
    ]
    address: "0xe24dBBEd860aD4b71cf69ED9c0AcEeeDcA99f934", 
    filter: {
      indexed: [{
        user: "0xf02e4e64f9f7fe6e461f136d248baab3c1f2a11e"
      }], 
      order: "desc", 
      limit: 200,
    range: {
      from: "06/01/2019"
    	to: "07/30/2019"
    },
    }) {
    meta{
      txID
    }
    txDate
    logName
    timestamp
    address
    decoded
  }
}
```
Author By @molekilla and Ruben Guevara

File: applications/solido.md
Comment (molekilla) [reply applications/solido.md:120]: any rosetta openapi implementation (btc, ethereum, vechain, kadena, ...)

File: applications/solido.md
Comment (alfredolopez80) [reply applications/solido.md]: ready

File: applications/solido.md
Comment (alfredolopez80) [reply applications/solido.md]: ready

File: applications/solido.md
Comment (mmagician) [reply applications/solido.md:120]: In your document you should specify which **concrete** chains/parachains will be available for interaction at the time of delivery of your milestone. That forms part of the contract and will later allow us to check whether you fulfilled the necessary conditions.

**Comment by molekilla** [comment]
Must Haves
- Moonbeam
- Plasm
- Edgeware
- PolkadotJS layer abstraction

Cheers

File: applications/solido.md
Comment (molekilla) [reply applications/solido.md:120]: @mmagician @Noc2 let us know anything left. right now api is for dapp development, not sure if relay chains are required

Comment (molekilla) [reply applications/solido.md:120]: eg paid needs to be migrated to Plasm, we're going to use Chainsafe to bridge to ethereum and Plasm with ink!, without Solido Substrate, because there is no universal adapter, we'll need to migrate ethereum dapp specific code to PolkadotJS. If we had tools like the proposed dapp development will be more faster to deliver

Comment (molekilla) [reply applications/solido.md:120]: aditionally, I had totally forgot, there are only two known ways to create native mobile dapps: using native wallet technology like Trust Wallet Core or use GraphQL.

With Solido Substrate, which is graphql client, both querying and sending signed transactions are possible without having to build complex wallet logic, only requirement is that the wallet manages the keys in a secure way and is able to sign tx for a specific chain

Comment (molekilla) [reply applications/solido.md:120]: @Noc2 @enfipy @trevelyan ^^^^^

**Comment by Noc2** [comment]
Sorry for the late reply. It seems so far the committee is not convinced by your application. The main reason is that we have already similar solutions and other big names in the ethereum ecosystem are planning to support polkadot soon. Could you point out the difference and the advantages of your solution (see my previous questions).

>  Are you aware of [Substrate Graph](https://github.com/playzero/substrate-graph) and especially [Hydra](https://github.com/Joystream/hydra). How is your solution different? And generally what’s your focus? (smart contracts in ink! or runtimes, validators or users, etc.) Or do you want to support everything?

Also could integrate the must-haves into the application md file?

**Comment by molekilla** [comment]
> Sorry for the late reply. It seems so far the committee is not convinced by your application. The main reason is that we have already similar solutions and other big names in the ethereum ecosystem are planning to support polkadot soon. Could you point out the difference and the advantages of your solution (see my previous questions).
> 
> > Are you aware of [Substrate Graph](https://github.com/playzero/substrate-graph) and especially [Hydra](https://github.com/Joystream/hydra). How is your solution different? And generally what’s your focus? (smart contracts in ink! or runtimes, validators or users, etc.) Or do you want to support everything?
> 
> Also could integrate the must-haves into the application md file?

Hi David

Assume for a moment that rosetta integration is overkill and that we need for some use cases a de facto standard to query and mutate transactions.

What and why would a dapp developer need that power?

- Substrate Graph and Hydra are TheGraph clones, eg they index contracts
- Our competitioin is really Web3Api.dev but they are starting from 0, we have Solido experience while working with Vechain since 2018

Certain scenarios, mainly native wallet and backend development, because of constraints, requires non human signing for example, in this cases best practice is to use TrustWallet Core as a tx signer and use eg Geth built in GraphQL endpoint

Either way, I'd  still develop this project because is a time saver 

thanks
will add must haves

**Comment by mmagician** [comment]
@molekilla Did you want to add some extra details? Otherwise, I'll close this PR shortly as it hasn't received any approvals.

**Comment by rubengue** [comment]
Hi @mmagician 

What if we instead of doing a client RPC to Rosseta, work on implementing the actual missing ingredient, which is a substrate api Sidecar for GraphQL. That will allow us to have parity with Ethereum, which has json RPC, WebSockets and GraphQL. We have the experience, becouse we created the only GraphQL endpoint for the VeChain Blockchain.

### References:

- https://github.com/PAIDNetwork/substrate-api-graphql/blob/main/README.md

- https://github.com/decent-bet/thor-connexql.git

Please let us know, me or  @molekilla, if there is any further questions. Thanks very much.

**Comment by mmagician** [comment]
Please have a look at a similar idea & comments for developing with GraphQL: https://github.com/w3f/Open-Grants-Program/pull/41

**Comment by rubengue** [comment]
https://eips.ethereum.org/EIPS/eip-1767

Hi guys, please read and comment//

Thanks very much

**Comment by mmagician** [comment]
@rubengue  I'm not sure what to make of your link. As already mentioned in https://github.com/w3f/Open-Grants-Program/pull/41, we wouldn't support another GraphQL project at the moment. I'm closing the PR, since it's been open for over a month without any progress in getting the approvals from the Grants Committee (+ earlier incomplete applications).

**Pull request closed without merging**
